lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -243,6 +332,72 @@ void FIRRTLModuleLowering::runOnOperation() {
   circuit.erase();
 }
 
+void FIRRTLModuleLowering::lowerMemoryDecls(const std::set<FirMemory> &mems,
+                                            Block *top,
+                                            CircuitLoweringState &state) {
+  auto b =
+      ImplicitLocOpBuilder::atBlockBegin(UnknownLoc::get(&getContext()), top);
+  std::array<StringRef, 6> schemaFields = {"depth",         "numReadPorts",
+                                           "numWritePorts", "readLatency",
+                                           "writeLatency",  "width"};
+  auto schemaFieldsAttr = b.getStrArrayAttr(schemaFields);
+  auto schema = b.create<rtl::RTLGeneratorSchemaOp>(
+      "FIRRTLMem", "FIRRTL_Memory", schemaFieldsAttr);
+  auto memorySchema = b.getSymbolRefAttr(schema);
+
+  SmallVector<rtl::ModulePortInfo> ports;
+  size_t inputPin = 0;
+  size_t outputPin = 0;
+
+  Type b1Type = IntegerType::get(&getContext(), 1);
+
+  auto makePortCommon = [&](StringRef prefix, Twine i, Type bAddrType) {
+    ports.push_back({b.getStringAttr((prefix + Twine("_clock_") + i).str()),
+                     rtl::INPUT, b1Type, inputPin++});
+    ports.push_back({b.getStringAttr((prefix + Twine("_en_") + i).str()),
+                     rtl::INPUT, b1Type, inputPin++});
+    ports.push_back({b.getStringAttr((prefix + Twine("_addr_") + i).str()),
+                     rtl::INPUT, bAddrType, inputPin++});
+  };
+
+  for (auto &mem : mems) {
+    Type bDataType;
+    if (mem.dataWidth > 0)
+      bDataType = IntegerType::get(&getContext(), mem.dataWidth);
```

Comment:
This doesn't handle the `mem.dataWidth == 0` case. Is that okay, e.g., have we fully removed zero-width memories before this?

Reply:
It does, but IntegerType can't be zero.  The data ports are elided for zero, so we don't need the type.